### PR TITLE
Add room support

### PR DESCRIPTION
**Pull request by scottinet**
Until now, implemented protocols in Kuzzle are able to dispatch messages to specific rooms, allowing client to listen to specific messages (AMQP, Socket.IO, MQTT, etc).

This is not the case with pure WebSockets, as this protocol do not feature such a functionality.

The purpose of this PR is to systematically add a `room` attribute to messages, allowing clients to dispatch incoming payloads themselves.

# Changelog

* add a `room` attribute to all outgoing messages
* update README file with a new `How to use` paragraph, explaining how to properly dispatch messages client-side
* update unit tests to check that the `room` attribute is correctly added
* heavy performances improvements
* bugfix: a connection close event now properly remove associated channels
* added eslint to "npm test" runs

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -45,7 +44,7 @@ function WebsocketProtocol () {
   };
 
   this.onServerError = function (error) {
-    console.error('An error has occured:', error);
+    console.error('[plugin-websocket] A error has occured:', error);
```
Comment (ballinette) [inline lib/index.js]: isn't it "A**n** error" ?

**Comment by ballinette** [comment]
+1

File: lib/index.js (hunk 1 above)
Comment (scottinet) [inline lib/index.js]: Oopsie.

**Comment by dbengsch** [comment]
+1

**Comment by scottinet** [comment]
Urgh. We have a memleak in kuzzle-proxy and preliminary analysis shows that it may come from this websocket plugin.

So... please don't merge this PR just yet. :-(

**Comment by scottinet** [comment]
We should be good, now. 
@ballinette , @dbengsch : you may want to review this PR once again, as I've made major changes since your approvals.

**Comment by scottinet** [comment]
Oooook. I know it's a bit weird, but once again, don't merge this one just yet. :-(

There is still a memleak, but this time, it appears to be with how NodeJS handles sockets. I'm trying to check if we need to throttle sends or if it's just V8's garbage collector who is too lazy to clean allocated memory. Which seems weird, because we can go up to 2.7GB allocated memory with the current implementation (and waaaay down with some throttling system).

File: LICENSE [hunk 2]
```diff
@@ -186,7 +186,7 @@
       same "printed page" as the copyright notice for easier
       identification within third-party archives.
 
-   Copyright 2016 Kuzzle
+   Copyright 2015 Kaliop SAS
```
Comment (benoitvidis) [inline LICENSE]: nitpicking: should be 2016

**Comment by benoitvidis** [comment]
Curious about getting the memleak final word but till then, lgtm.

**Comment by ballinette** [comment]
still OK for me

**Comment by ballinette** [comment]
(NB: shouldn't we add a label "WIP" if this PR is still not ready to merge ?)

**Comment by scottinet** [comment]
I'm too lazy to create the tag on this project. :p

**Comment by scottinet** [comment]
@benoitvidis > well... it's hard to know exactly what's going on. What I can say is, even by calling the GC manually, V8 is very slow to free allocated memory, and a lot is still there after huge volumes of activity.

Moreover, it seems to be closely related to how sockets are under pressure or note. When I throttle sockets even a little, memory allocation is way more reasonable (I'm speaking 90MB instead of 2.7GB here).

The latest changes introduce a change in how Kuzzle sends notifications. Instead of sending 1 notification per channel, which was fine when there was no proxy in front of servers, Kuzzle now sends 1 notification with a list of channels to notify. 
This reduce network usage A LOT, and memory allocation in this plugin now don't exceed 300MB. 

I need to dig deeper in Node forums and issues, or even in V8 ones, in order to get a better idea of what's going on.

For now... this is the final version of this plugin. I swear. 
In addition of the multichannels notifications I mentioned earlier, I've also fixed a bug where a client socket could be used while it's closing, and I reduced usage of `Object.keys(this.channels)` which could be a bit slow on this potentially very large object.

This PR can, at last, be reviewed.

**Comment by dbengsch** [comment]
+1 (again after all the changes)

**Pull request merged**